sql/memo/memo.go

```diff
@@ -326,6 +305,20 @@ func (m *Memo) optimizeMemoGroup(grp *ExprGroup) error {
 	}
 	var err error
 	n := grp.First
+	if _, ok := n.(SourceRel); ok {
```

Comment:
This block needs a comment, and maybe should outlined. I eventually figured out that we want to choose IndexScans over other sourceRel exoressions, but it clutters the rest of the function.

Reply:
Yeah this points out a key limitation of the way we order the search. Ideally we would generate plan options bottom-up the way we do now, but with the full plan materialized for each node. So we would have a generator capable of keeping track of which plans to check, but the forest of plans and groups should maybe be abstracted from the actual costing logic. The memo representation of inner join would reference indexScans, and a lookupJoin in the same group's LHS can be indexScan/tableScan and RHS be tableScans. The physical operators can share group properties for cardinality/FD tracking, but get rid of the indirection and weird dependencies going on right now.